reconcile/ocm_machine_pools.py

```diff
@@ -1,137 +1,332 @@
+    def create_from_gql(cls, pool: ClusterMachinePoolV1, cluster: str):
+        return cls(
+            id=pool.q_id,
+            replicas=pool.replicas,
+            instance_type=pool.instance_type,
+            taints=pool.taints,
+            labels=pool.labels,
+            cluster=cluster,
+        )
+
+
+class AWSNodePool(BaseModel):
+    instance_type: str
+
+
+class NodePool(AbstractPool):
+    aws_node_pool: AWSNodePool
+    subnet: Optional[str]
 
-    current_state = []
+    def delete(self, ocm: OCM) -> None:
+        ocm.delete_node_pool(self.cluster, self.dict(by_alias=True))
+
+    def create(self, ocm: OCM) -> None:
+        ocm.create_node_pool(self.cluster, self.dict(by_alias=True))
+
+    def update(self, ocm: OCM) -> None:
+        ocm.update_node_pool(self.cluster, self.dict(by_alias=True))
+
+    def has_diff(self, pool: ClusterMachinePoolV1) -> bool:
+        return (
+            self.replicas != pool.replicas
+            or self.taints != pool.taints
+            or self.labels != pool.labels
+        )
+
+    def invalid_diff(self, pool: ClusterMachinePoolV1) -> Optional[str]:
+        if self.aws_node_pool.instance_type != pool.instance_type:
+            return "instance_type"
+        if self.subnet != pool.subnet:
+            return "subnet"
+
+    @classmethod
+    def create_from_gql(cls, pool: ClusterMachinePoolV1, cluster: str):
+        return cls(
+            id=pool.q_id,
+            replicas=pool.replicas,
+            aws_node_pool=AWSNodePool(
+                instance_type=pool.instance_type,
+            ),
+            taints=pool.taints,
+            labels=pool.labels,
+            subnet=pool.subnet,
+            cluster=cluster,
+        )
+
+
+class PoolHandler(BaseModel):
+    action: str
+    pool: AbstractPool
+
+    def act(self, dry_run: bool, ocm: OCM) -> None:
+        logging.info(f"{self.action} {self.pool.dict(by_alias=True)}")
+        if dry_run:
+            return
+
+        if not self.action:
+            pass
+        elif self.action == "delete":
+            self.pool.delete(ocm)
+        elif self.action == "create":
+            self.pool.create(ocm)
+
+
+def fetch_current_state(
+    ocm_map: OCMMap,
+    clusters: Iterable[ClusterV1],
+) -> Mapping[str, Mapping[str, list[AbstractPool]]]:
+    current_state = {}
     for cluster in clusters:
-        cluster_name = cluster["name"]
-        ocm = ocm_map.get(cluster_name)
-        machine_pools = ocm.get_machine_pools(cluster_name)
-        for machine_pool in machine_pools:
-            machine_pool["cluster"] = cluster_name
-            current_state.append(machine_pool)
+        current_state[cluster.name] = {
+            "machine_pools": [],
+        }
+        ocm = ocm_map.get(cluster.name)
+        if cluster.spec.hypershift:
+            machine_pools = ocm.get_node_pools(cluster.name)
+            for machine_pool in machine_pools:
+                current_state[cluster.name]["machine_pools"].append(
```

Comment:
so we store `node_pools` in our `machine_pools` list for our current state?

Reply:
Actually not required, removing the submapping